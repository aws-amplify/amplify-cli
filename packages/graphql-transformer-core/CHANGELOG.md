# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="1.0.28"></a>
## [1.0.28](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.28-beta.0...graphql-transformer-core@1.0.28) (2018-10-18)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.28-beta.0"></a>
## [1.0.28-beta.0](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.12...graphql-transformer-core@1.0.28-beta.0) (2018-10-12)


### Bug Fixes

* **graphql-transformer-core:** Fix Readme.md docs for the `[@auth](https://github.com/auth)` directive ([db6ff7a](https://github.com/aws-amplify/amplify-cli/commit/db6ff7a))




<a name="1.0.12"></a>
## [1.0.12](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.11...graphql-transformer-core@1.0.12) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.11"></a>
## [1.0.11](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.10...graphql-transformer-core@1.0.11) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.10"></a>
## [1.0.10](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.9...graphql-transformer-core@1.0.10) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.9"></a>
## [1.0.9](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.8...graphql-transformer-core@1.0.9) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.8"></a>
## [1.0.8](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.7...graphql-transformer-core@1.0.8) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.7"></a>
## [1.0.7](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.6...graphql-transformer-core@1.0.7) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.6"></a>
## [1.0.6](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.5...graphql-transformer-core@1.0.6) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.5"></a>
## [1.0.5](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.4...graphql-transformer-core@1.0.5) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.4"></a>
## 1.0.4 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.3"></a>
## 1.0.3 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.2"></a>
## 1.0.2 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.1"></a>
## 1.0.1 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core
