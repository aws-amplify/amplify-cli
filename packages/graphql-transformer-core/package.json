{
  "name": "graphql-transformer-core",
  "version": "6.19.4",
  "description": "A framework to transform from GraphQL SDL to AWS CloudFormation.",
  "repository": {
    "type": "git",
    "url": "https://github.com/aws-amplify/amplify-cli.git",
    "directory": "packages/graphql-transformer-core"
  },
  "author": "Amazon Web Services",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "keywords": [
    "graphql",
    "cloudformation",
    "aws"
  ],
  "scripts": {
    "test": "jest",
    "test-ci": "jest --ci -i",
    "build": "tsc",
    "watch": "tsc -w",
    "clean": "rimraf ./lib"
  },
  "dependencies": {
    "cloudform-types": "^4.2.0",
    "deep-diff": "^1.0.2",
    "fs-extra": "^8.1.0",
    "glob": "^7.1.6",
    "graphql": "^14.5.8",
    "graphql-transformer-common": "4.17.4"
  },
  "devDependencies": {
    "@types/fs-extra": "^8.0.1",
    "@types/node": "^10.17.13"
  },
  "jest": {
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "testURL": "http://localhost",
    "testRegex": "(src/__tests__/.*|(\\.|/)(test|spec))\\.(jsx?|tsx?)$",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ],
    "collectCoverage": true
  }
}
