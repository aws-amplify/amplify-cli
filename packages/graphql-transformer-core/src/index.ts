import './polyfills/Object.assign';
import { print } from 'graphql';
import { TransformerContext } from './TransformerContext';
import { Transformer } from './Transformer';
import { ITransformer } from './ITransformer';
import { GraphQLTransform } from './GraphQLTransform';
import { collectDirectiveNames, collectDirectivesByTypeNames } from './collectDirectives';
import { stripDirectives } from './stripDirectives';
import { DeploymentResources } from './DeploymentResources';
import {
  buildProject as buildAPIProject,
  uploadDeployment as uploadAPIProject,
  migrateAPIProject,
  revertAPIMigration,
} from './util/amplifyUtils';
import {
  readSchema as readProjectSchema,
  loadProject as readProjectConfiguration,
  loadConfig as readTransformerConfiguration,
  writeConfig as writeTransformerConfiguration,
  TRANSFORM_CONFIG_FILE_NAME,
  TRANSFORM_BASE_VERSION,
  TRANSFORM_CURRENT_VERSION,
  TransformConfig,
  SyncConfig,
} from './util/transformConfig';
import { EXTRA_DIRECTIVES_DOCUMENT } from './validation';
export * from './errors';
export * from './util';

/**
 * Returns the extra set of directives that are supported by AppSync service
 */
export function getAppSyncServiceExtraDirectives(): string {
  return print(EXTRA_DIRECTIVES_DOCUMENT);
}

export {
  GraphQLTransform,
  TransformConfig,
  TransformerContext,
  Transformer,
  ITransformer,
  collectDirectiveNames,
  collectDirectivesByTypeNames,
  stripDirectives,
  buildAPIProject,
  migrateAPIProject,
  uploadAPIProject,
  readProjectSchema,
  readProjectConfiguration,
  readTransformerConfiguration,
  writeTransformerConfiguration,
  revertAPIMigration,
  TRANSFORM_CONFIG_FILE_NAME,
  TRANSFORM_BASE_VERSION,
  TRANSFORM_CURRENT_VERSION,
  SyncConfig,
  DeploymentResources,
};
