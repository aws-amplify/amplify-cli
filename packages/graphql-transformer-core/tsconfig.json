{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "strict": false, // TODO enable
    "rootDir": "src",
    "outDir": "lib"
  },
  "references": [
    {"path": "../graphql-transformer-common"}
  ]
}
