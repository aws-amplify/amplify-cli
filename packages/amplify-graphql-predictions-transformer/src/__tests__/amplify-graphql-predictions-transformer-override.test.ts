import { GraphQLTransform } from '@aws-amplify/graphql-transformer-core';
import * as path from 'path';
import { PredictionsTransformer } from '..';

test('it generates resources with overrides', () => {
  const validSchema = /* GraphQL */ `
    type Query {
      speakTranslatedIdentifiedText: String @predictions(actions: [identifyText, translateText, convertTextToSpeech])
      speakTranslatedLabelText: String @predictions(actions: [identifyLabels, translateText, convertTextToSpeech])
    }
  `;
  const transformer = new GraphQLTransform({
    transformers: [new PredictionsTransformer({ bucketName: 'myStorage${hash}-${env}' })],
    overrideConfig: {
      overrideDir: path.join(__dirname, 'overrides'),
      overrideFlag: true,
      resourceName: 'myResource',
    },
  });

  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.stacks).toMatchSnapshot();
});
