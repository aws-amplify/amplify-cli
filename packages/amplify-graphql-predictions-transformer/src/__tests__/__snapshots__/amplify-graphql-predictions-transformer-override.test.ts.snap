// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it generates resources with overrides 1`] = `
Object {
  "PredictionsDirectiveStack": Object {
    "Conditions": Object {
      "HasEnvironmentParameter": Object {
        "Fn::Not": Array [
          Object {
            "Fn::Equals": Array [
              Object {
                "Ref": "referencetotransformerrootstackenv10C5A902Ref",
              },
              "NONE",
            ],
          },
        ],
      },
    },
    "Parameters": Object {
      "referencetotransformerrootstackGraphQLAPI20497F53ApiId": Object {
        "Type": "String",
      },
      "referencetotransformerrootstackS3DeploymentBucket7592718ARef": Object {
        "Type": "String",
      },
      "referencetotransformerrootstackS3DeploymentRootKeyA71EA735Ref": Object {
        "Type": "String",
      },
      "referencetotransformerrootstackenv10C5A902Ref": Object {
        "Type": "String",
      },
    },
    "Resources": Object {
      "LambdaDataSource": Object {
        "Properties": Object {
          "ApiId": Object {
            "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
          },
          "LambdaConfig": Object {
            "LambdaFunctionArn": Object {
              "Fn::GetAtt": Array [
                "predictionsLambda8B880A74",
                "Arn",
              ],
            },
          },
          "Name": "LambdaDataSource",
          "ServiceRoleArn": Object {
            "Fn::GetAtt": Array [
              "predictionsIAMRole5BB74A99",
              "Arn",
            ],
          },
          "Type": "AWS_LAMBDA",
        },
        "Type": "AWS::AppSync::DataSource",
      },
      "LambdaDataSourceServiceRole81BB3362": Object {
        "Properties": Object {
          "AssumeRolePolicyDocument": Object {
            "Statement": Array [
              Object {
                "Action": "sts:AssumeRole",
                "Effect": "Allow",
                "Principal": Object {
                  "Service": "appsync.amazonaws.com",
                },
              },
            ],
            "Version": "2012-10-17",
          },
        },
        "Type": "AWS::IAM::Role",
      },
      "LambdaDataSourceServiceRoleDefaultPolicy00CC76B1": Object {
        "Properties": Object {
          "PolicyDocument": Object {
            "Statement": Array [
              Object {
                "Action": "lambda:InvokeFunction",
                "Effect": "Allow",
                "Resource": Object {
                  "Fn::GetAtt": Array [
                    "predictionsLambda8B880A74",
                    "Arn",
                  ],
                },
              },
            ],
            "Version": "2012-10-17",
          },
          "PolicyName": "LambdaDataSourceServiceRoleDefaultPolicy00CC76B1",
          "Roles": Array [
            Object {
              "Ref": "LambdaDataSourceServiceRole81BB3362",
            },
          ],
        },
        "Type": "AWS::IAM::Policy",
      },
      "PredictionsLambdaAccess4D9DA807": Object {
        "Properties": Object {
          "PolicyDocument": Object {
            "Statement": Array [
              Object {
                "Action": "lambda:InvokeFunction",
                "Effect": "Allow",
                "Resource": Object {
                  "Fn::GetAtt": Array [
                    "predictionsLambda8B880A74",
                    "Arn",
                  ],
                },
              },
            ],
            "Version": "2012-10-17",
          },
          "PolicyName": "PredictionsLambdaAccess4D9DA807",
          "Roles": Array [
            Object {
              "Ref": "predictionsIAMRole5BB74A99",
            },
          ],
        },
        "Type": "AWS::IAM::Policy",
      },
      "PredictionsStorageAccess68CD5140": Object {
        "Properties": Object {
          "PolicyDocument": Object {
            "Statement": Array [
              Object {
                "Action": "s3:GetObject",
                "Effect": "Allow",
                "Resource": Object {
                  "Fn::If": Array [
                    "HasEnvironmentParameter",
                    Object {
                      "Fn::Sub": Array [
                        "arn:aws:s3:::myStorage\${hash}-\${env}/public/*",
                        Object {
                          "env": Object {
                            "Ref": "referencetotransformerrootstackenv10C5A902Ref",
                          },
                          "hash": Object {
                            "Fn::Select": Array [
                              3,
                              Object {
                                "Fn::Split": Array [
                                  "-",
                                  Object {
                                    "Ref": "AWS::StackName",
                                  },
                                ],
                              },
                            ],
                          },
                        },
                      ],
                    },
                    Object {
                      "Fn::Sub": Array [
                        "arn:aws:s3:::myStorage\${hash}/public/*",
                        Object {
                          "hash": Object {
                            "Fn::Select": Array [
                              3,
                              Object {
                                "Fn::Split": Array [
                                  "-",
                                  Object {
                                    "Ref": "AWS::StackName",
                                  },
                                ],
                              },
                            ],
                          },
                        },
                      ],
                    },
                  ],
                },
              },
            ],
            "Version": "2012-10-17",
          },
          "PolicyName": "PredictionsStorageAccess68CD5140",
          "Roles": Array [
            Object {
              "Ref": "predictionsIAMRole5BB74A99",
            },
          ],
        },
        "Type": "AWS::IAM::Policy",
      },
      "QueryspeakTranslatedIdentifiedTextResolver": Object {
        "Properties": Object {
          "ApiId": Object {
            "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
          },
          "FieldName": "speakTranslatedIdentifiedText",
          "Kind": "PIPELINE",
          "PipelineConfig": Object {
            "Functions": Array [
              Object {
                "Fn::GetAtt": Array [
                  "identifyTextFunctionidentifyTextFunctionAppSyncFunction7877885A",
                  "FunctionId",
                ],
              },
              Object {
                "Fn::GetAtt": Array [
                  "translateTextFunctiontranslateTextFunctionAppSyncFunction77C9D7A9",
                  "FunctionId",
                ],
              },
              Object {
                "Fn::GetAtt": Array [
                  "convertTextToSpeechFunctionconvertTextToSpeechFunctionAppSyncFunction15BC20B8",
                  "FunctionId",
                ],
              },
            ],
          },
          "RequestMappingTemplate": Object {
            "Fn::Join": Array [
              "
",
              Array [
                Object {
                  "Fn::If": Array [
                    "HasEnvironmentParameter",
                    Object {
                      "Fn::Sub": Array [
                        "$util.qr($ctx.stash.put(\\"s3Bucket\\", \\"myStorage\${hash}-\${env}\\"))",
                        Object {
                          "env": Object {
                            "Ref": "referencetotransformerrootstackenv10C5A902Ref",
                          },
                          "hash": Object {
                            "Fn::Select": Array [
                              3,
                              Object {
                                "Fn::Split": Array [
                                  "-",
                                  Object {
                                    "Ref": "AWS::StackName",
                                  },
                                ],
                              },
                            ],
                          },
                        },
                      ],
                    },
                    Object {
                      "Fn::Sub": Array [
                        "$util.qr($ctx.stash.put(\\"s3Bucket\\", \\"myStorage\${hash}\\"))",
                        Object {
                          "hash": Object {
                            "Fn::Select": Array [
                              3,
                              Object {
                                "Fn::Split": Array [
                                  "-",
                                  Object {
                                    "Ref": "AWS::StackName",
                                  },
                                ],
                              },
                            ],
                          },
                        },
                      ],
                    },
                  ],
                },
                "$util.qr($ctx.stash.put(\\"isList\\", false))
{}",
              ],
            ],
          },
          "ResponseMappingTemplate": "## If the result is a list return the result as a list **
#if( $ctx.stash.get(\\"isList\\") )
  #set( $result = $ctx.result.split(\\"[ ,]+\\") )
  $util.toJson($result)
#else
  $util.toJson($ctx.result)
#end",
          "TypeName": "Query",
        },
        "Type": "AWS::AppSync::Resolver",
      },
      "QueryspeakTranslatedLabelTextResolver": Object {
        "Properties": Object {
          "ApiId": Object {
            "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
          },
          "FieldName": "speakTranslatedLabelText",
          "Kind": "PIPELINE",
          "PipelineConfig": Object {
            "Functions": Array [
              Object {
                "Fn::GetAtt": Array [
                  "identifyLabelsFunctionidentifyLabelsFunctionAppSyncFunction7C2A1E61",
                  "FunctionId",
                ],
              },
              Object {
                "Fn::GetAtt": Array [
                  "translateTextFunctiontranslateTextFunctionAppSyncFunction77C9D7A9",
                  "FunctionId",
                ],
              },
              Object {
                "Fn::GetAtt": Array [
                  "convertTextToSpeechFunctionconvertTextToSpeechFunctionAppSyncFunction15BC20B8",
                  "FunctionId",
                ],
              },
            ],
          },
          "RequestMappingTemplate": "mockTeplate",
          "ResponseMappingTemplate": "## If the result is a list return the result as a list **
#if( $ctx.stash.get(\\"isList\\") )
  #set( $result = $ctx.result.split(\\"[ ,]+\\") )
  $util.toJson($result)
#else
  $util.toJson($ctx.result)
#end",
          "TypeName": "Query",
        },
        "Type": "AWS::AppSync::Resolver",
      },
      "RekognitionDataSource": Object {
        "Properties": Object {
          "ApiId": Object {
            "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
          },
          "HttpConfig": Object {
            "AuthorizationConfig": Object {
              "AuthorizationType": "AWS_IAM",
              "AwsIamConfig": Object {
                "SigningRegion": Object {
                  "Fn::Sub": "\${AWS::Region}",
                },
                "SigningServiceName": "rekognition",
              },
            },
            "Endpoint": Object {
              "Fn::Sub": "https://rekognition.\${AWS::Region}.amazonaws.com",
            },
          },
          "Name": "RekognitionDataSource",
          "ServiceRoleArn": Object {
            "Fn::GetAtt": Array [
              "predictionsIAMRole5BB74A99",
              "Arn",
            ],
          },
          "Type": "HTTP",
        },
        "Type": "AWS::AppSync::DataSource",
      },
      "RekognitionDataSourceServiceRole3628B5DB": Object {
        "Properties": Object {
          "AssumeRolePolicyDocument": Object {
            "Statement": Array [
              Object {
                "Action": "sts:AssumeRole",
                "Effect": "Allow",
                "Principal": Object {
                  "Service": "appsync.amazonaws.com",
                },
              },
            ],
            "Version": "2012-10-17",
          },
        },
        "Type": "AWS::IAM::Role",
      },
      "TranslateDataSource": Object {
        "Properties": Object {
          "ApiId": Object {
            "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
          },
          "HttpConfig": Object {
            "AuthorizationConfig": Object {
              "AuthorizationType": "AWS_IAM",
              "AwsIamConfig": Object {
                "SigningRegion": Object {
                  "Fn::Sub": "\${AWS::Region}",
                },
                "SigningServiceName": "translate",
              },
            },
            "Endpoint": Object {
              "Fn::Sub": "https://translate.\${AWS::Region}.amazonaws.com",
            },
          },
          "Name": "TranslateDataSource",
          "ServiceRoleArn": "mockArn",
          "Type": "HTTP",
        },
        "Type": "AWS::AppSync::DataSource",
      },
      "TranslateDataSourceServiceRole2FE48B83": Object {
        "Properties": Object {
          "AssumeRolePolicyDocument": Object {
            "Statement": Array [
              Object {
                "Action": "sts:AssumeRole",
                "Effect": "Allow",
                "Principal": Object {
                  "Service": "appsync.amazonaws.com",
                },
              },
            ],
            "Version": "2012-10-17",
          },
        },
        "Type": "AWS::IAM::Role",
      },
      "convertTextToSpeechFunctionconvertTextToSpeechFunctionAppSyncFunction15BC20B8": Object {
        "DependsOn": Array [
          "LambdaDataSource",
        ],
        "Properties": Object {
          "ApiId": Object {
            "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
          },
          "DataSourceName": Object {
            "Fn::GetAtt": Array [
              "LambdaDataSource",
              "Name",
            ],
          },
          "FunctionVersion": "2018-05-29",
          "Name": "convertTextToSpeechFunction",
          "RequestMappingTemplate": "#set( $bucketName = $ctx.stash.get(\\"s3Bucket\\") )
$util.qr($ctx.stash.put(\\"isList\\", false))
#set( $text = $util.defaultIfNull($ctx.args.input.convertTextToSpeech.text, $ctx.prev.result) )
{
  \\"version\\": \\"2018-05-29\\",
  \\"operation\\": \\"Invoke\\",
  \\"payload\\":   $util.toJson({
  \\"uuid\\": \\"$util.autoId()\\",
  \\"action\\": \\"convertTextToSpeech\\",
  \\"voiceID\\": $ctx.args.input.convertTextToSpeech.voiceID,
  \\"text\\": $text
})
}",
          "ResponseMappingTemplate": "#if( $ctx.error )
$util.error($ctx.error.message, $ctx.error.type)
#end
#set( $response = $util.parseJson($ctx.result) )
$util.toJson($ctx.result.url)",
        },
        "Type": "AWS::AppSync::FunctionConfiguration",
      },
      "identifyLabelsAccess7BFAF94B": Object {
        "Properties": Object {
          "PolicyDocument": Object {
            "Statement": Array [
              Object {
                "Action": "rekognition:DetectLabels",
                "Effect": "Allow",
                "Resource": "*",
              },
            ],
            "Version": "2012-10-17",
          },
          "PolicyName": "identifyLabelsAccess7BFAF94B",
          "Roles": Array [
            Object {
              "Ref": "predictionsIAMRole5BB74A99",
            },
          ],
        },
        "Type": "AWS::IAM::Policy",
      },
      "identifyLabelsFunctionidentifyLabelsFunctionAppSyncFunction7C2A1E61": Object {
        "DependsOn": Array [
          "RekognitionDataSource",
        ],
        "Properties": Object {
          "ApiId": Object {
            "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
          },
          "DataSourceName": Object {
            "Fn::GetAtt": Array [
              "RekognitionDataSource",
              "Name",
            ],
          },
          "FunctionVersion": "2018-05-29",
          "Name": "identifyLabelsFunction",
          "RequestMappingTemplate": "#set( $bucketName = $ctx.stash.get(\\"s3Bucket\\") )
#set( $identifyLabelKey = $ctx.args.input.identifyLabels.key )
$util.qr($ctx.stash.put(\\"isList\\", true))
#set( $identifyLabelsPayload = {
  \\"version\\": \\"2018-05-29\\",
  \\"method\\": \\"POST\\",
  \\"resourcePath\\": \\"/\\",
  \\"params\\": {
      \\"body\\": {
          \\"Image\\": {
              \\"S3Object\\": {
                  \\"Bucket\\": \\"$bucketName\\",
                  \\"Name\\": \\"public/$identifyLabelKey\\"
        }
      },
          \\"MaxLabels\\": 10,
          \\"MinConfidence\\": 55
    },
      \\"headers\\": {
          \\"Content-Type\\": \\"application/x-amz-json-1.1\\",
          \\"X-Amz-Target\\": \\"RekognitionService.DetectLabels\\"
    }
  }
} )
$util.toJson($identifyLabelsPayload)",
          "ResponseMappingTemplate": "#if( $ctx.error )
$util.error($ctx.error.message)
#end
#if( $ctx.result.statusCode == 200 )
  #set( $labels = \\"\\" )
  #set( $result = $util.parseJson($ctx.result.body) )
  #foreach( $label in $result.Labels )
    #set( $labels = \\"$labels$label.Name, \\" )
  #end
  $util.toJson($labels.replaceAll(\\", $\\", \\"\\"))
#else
$util.error($ctx.result.body)
#end",
        },
        "Type": "AWS::AppSync::FunctionConfiguration",
      },
      "identifyTextAccess5A2D7702": Object {
        "Properties": Object {
          "PolicyDocument": Object {
            "Statement": Array [
              Object {
                "Action": "rekognition:DetectText",
                "Effect": "Allow",
                "Resource": "*",
              },
            ],
            "Version": "2012-10-17",
          },
          "PolicyName": "identifyTextAccess5A2D7702",
          "Roles": Array [
            Object {
              "Ref": "predictionsIAMRole5BB74A99",
            },
          ],
        },
        "Type": "AWS::IAM::Policy",
      },
      "identifyTextFunctionidentifyTextFunctionAppSyncFunction7877885A": Object {
        "DependsOn": Array [
          "RekognitionDataSource",
        ],
        "Properties": Object {
          "ApiId": Object {
            "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
          },
          "DataSourceName": Object {
            "Fn::GetAtt": Array [
              "RekognitionDataSource",
              "Name",
            ],
          },
          "FunctionVersion": "2018-05-29",
          "Name": "identifyTextFunction",
          "RequestMappingTemplate": "#set( $bucketName = $ctx.stash.get(\\"s3Bucket\\") )
#set( $identityTextKey = $ctx.args.input.identifyText.key )
#set( $identifyTextPayload = {
  \\"version\\": \\"2018-05-29\\",
  \\"method\\": \\"POST\\",
  \\"resourcePath\\": \\"/\\",
  \\"params\\": {
      \\"body\\": {
          \\"Image\\": {
              \\"S3Object\\": {
                  \\"Bucket\\": \\"$bucketName\\",
                  \\"Name\\": \\"public/$identityTextKey\\"
        }
      }
    },
      \\"headers\\": {
          \\"Content-Type\\": \\"application/x-amz-json-1.1\\",
          \\"X-Amz-Target\\": \\"RekognitionService.DetectText\\"
    }
  }
} )
$util.toJson($identifyTextPayload)",
          "ResponseMappingTemplate": "#if( $ctx.error )
$util.error($ctx.error.message)
#end
#if( $ctx.result.statusCode == 200 )
  #set( $results = $util.parseJson($ctx.result.body) )
  #set( $finalResult = \\"\\" )
  #foreach( $item in $results.TextDetections )
    #if( $item.Type == \\"LINE\\" )
      #set( $finalResult = \\"$finalResult$item.DetectedText \\" )
    #end
  #end
$util.toJson($finalResult.trim())
#else
$utils.error($ctx.result.body)
#end",
        },
        "Type": "AWS::AppSync::FunctionConfiguration",
      },
      "predictionsIAMRole5BB74A99": Object {
        "Properties": Object {
          "AssumeRolePolicyDocument": Object {
            "Statement": Array [
              Object {
                "Action": "sts:AssumeRole",
                "Effect": "Allow",
                "Principal": Object {
                  "Service": "appsync.amazonaws.com",
                },
              },
            ],
            "Version": "2012-10-17",
          },
          "RoleName": Object {
            "Fn::If": Array [
              "HasEnvironmentParameter",
              Object {
                "Fn::Join": Array [
                  "-",
                  Array [
                    "predictionsIAMRole",
                    Object {
                      "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
                    },
                    Object {
                      "Ref": "referencetotransformerrootstackenv10C5A902Ref",
                    },
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "-",
                  Array [
                    "predictionsIAMRole",
                    Object {
                      "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
                    },
                  ],
                ],
              },
            ],
          },
        },
        "Type": "AWS::IAM::Role",
      },
      "predictionsLambda8B880A74": Object {
        "DependsOn": Array [
          "predictionsLambdaIAMRoleB8AC96D3",
        ],
        "Properties": Object {
          "Code": Object {
            "S3Bucket": Object {
              "Ref": "referencetotransformerrootstackS3DeploymentBucket7592718ARef",
            },
            "S3Key": Object {
              "Fn::Join": Array [
                "",
                Array [
                  Object {
                    "Ref": "referencetotransformerrootstackS3DeploymentRootKeyA71EA735Ref",
                  },
                  "/functions/predictionsLambdaFunction.zip",
                ],
              ],
            },
          },
          "Handler": "predictionsLambda.handler",
          "Role": Object {
            "Fn::GetAtt": Array [
              "predictionsLambdaIAMRoleB8AC96D3",
              "Arn",
            ],
          },
          "Runtime": "nodejs12.x",
          "Timeout": 60,
        },
        "Type": "AWS::Lambda::Function",
      },
      "predictionsLambdaIAMRoleB8AC96D3": Object {
        "Properties": Object {
          "AssumeRolePolicyDocument": Object {
            "Statement": Array [
              Object {
                "Action": "sts:AssumeRole",
                "Effect": "Allow",
                "Principal": Object {
                  "Service": "lambda.amazonaws.com",
                },
              },
            ],
            "Version": "2012-10-17",
          },
          "Policies": Array [
            Object {
              "PolicyDocument": Object {
                "Statement": Array [
                  Object {
                    "Action": "polly:SynthesizeSpeech",
                    "Effect": "Allow",
                    "Resource": "*",
                  },
                ],
                "Version": "2012-10-17",
              },
              "PolicyName": "PollyAccess",
            },
          ],
          "RoleName": Object {
            "Fn::If": Array [
              "HasEnvironmentParameter",
              Object {
                "Fn::Join": Array [
                  "-",
                  Array [
                    "predictionsLambdaIAMRole",
                    Object {
                      "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
                    },
                    Object {
                      "Ref": "referencetotransformerrootstackenv10C5A902Ref",
                    },
                  ],
                ],
              },
              Object {
                "Fn::Join": Array [
                  "-",
                  Array [
                    "predictionsLambdaIAMRole",
                    Object {
                      "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
                    },
                  ],
                ],
              },
            ],
          },
        },
        "Type": "AWS::IAM::Role",
      },
      "translateTextAccess74D3AE90": Object {
        "Properties": Object {
          "PolicyDocument": Object {
            "Statement": Array [
              Object {
                "Action": "translate:TranslateText",
                "Effect": "Allow",
                "Resource": "*",
              },
            ],
            "Version": "2012-10-17",
          },
          "PolicyName": "translateTextAccess74D3AE90",
          "Roles": Array [
            Object {
              "Ref": "predictionsIAMRole5BB74A99",
            },
          ],
        },
        "Type": "AWS::IAM::Policy",
      },
      "translateTextFunctiontranslateTextFunctionAppSyncFunction77C9D7A9": Object {
        "DependsOn": Array [
          "TranslateDataSource",
        ],
        "Properties": Object {
          "ApiId": Object {
            "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
          },
          "DataSourceName": Object {
            "Fn::GetAtt": Array [
              "TranslateDataSource",
              "Name",
            ],
          },
          "FunctionVersion": "2018-05-29",
          "Name": "translateTextFunction",
          "RequestMappingTemplate": "#set( $text = $util.defaultIfNull($ctx.args.input.translateText.text, $ctx.prev.result) )
#set( $translateTextPayload = {
  \\"version\\": \\"2018-05-29\\",
  \\"method\\": \\"POST\\",
  \\"resourcePath\\": \\"/\\",
  \\"params\\": {
      \\"body\\": {
          \\"SourceLanguageCode\\": $ctx.args.input.translateText.sourceLanguage,
          \\"TargetLanguageCode\\": $ctx.args.input.translateText.targetLanguage,
          \\"Text\\": $text
    },
      \\"headers\\": {
          \\"Content-Type\\": \\"application/x-amz-json-1.1\\",
          \\"X-Amz-Target\\": \\"AWSShineFrontendService_20170701.TranslateText\\"
    }
  }
} )
$util.toJson($translateTextPayload)",
          "ResponseMappingTemplate": "#if( $ctx.error )
$util.error($ctx.error.message)
#end
#if( $ctx.result.statusCode == 200 )
  #set( $result = $util.parseJson($ctx.result.body) )
$util.toJson($result.TranslatedText)
#else
$util.error($ctx.result.body)
#end",
        },
        "Type": "AWS::AppSync::FunctionConfiguration",
      },
    },
  },
}
`;
